js/src/main/scala/com/raquo/waypoint/Router.scala

```diff
@@ -333,8 +333,55 @@ class Router[BasePage](
   }
 
 }
-
 object Router {
+  /** The part of router that does not depend on the page type. */
+  trait Shared {
+    /** @see [[Router.replacePageTitle]] */
+    def replacePageTitle(title: String): Unit
+
+    /** @see [[Router.origin]] */
+    def origin: String
+  }
+
+  /** The contravariant side of the router. */
+  trait In[-BasePage] extends Shared {
+
+    /** @see [[Router.absoluteUrlForPage]] */
+    def absoluteUrlForPage(page: BasePage): String
+
+    /** @see [[Router.relativeUrlForPage]] */
+    def relativeUrlForPage(page: BasePage): String
+
+    /** @see [[Router.pushState]] */
+    def pushState(page: BasePage): Unit
+
+    /** @see [[Router.replaceState]] */
+    def replaceState(page: BasePage): Unit
+
+    /** @see [[Router.forcePage]] */
+    def forcePage(page: BasePage): Unit
+
+    /** @see [[Router.navigateTo]] */
+    def navigateTo(
+      page: BasePage,
+      replaceState: Boolean = false
+    ): Binder[Element]
+  }
+
+  /** The covariant side of the router. */
+  trait Out[+BasePage] extends Shared {
+    /** @see [[Router.currentPageSignal]] */
+    def currentPageSignal: StrictSignal[BasePage]
```

Comment:
Latest version LGTM except that `def currentPageSignal` and `def origin` both should be val-s. The origin is not supposed to be lazy or changeable, and `currentPageSignal` can't really be implemented as a `def` because it pretty much has to have side effects inside (so if we did, it would likely break) (also there should be no reason to implement it as a def).